[10.0][FIX] account_invoice_triple_discount: Fix dependency issue

acsonefho opened the pull request:
**Bug**
In some case, amounts are not corrects (subtotal etc).
It's because each iteration (of api.multi) could be done in different cache.
As we're updating cache in this case, it could be possible that another cache is used, so without edited values. So the amount is not the expected one.
To fix that, we just have to use api.one (like the original method in Odoo) to stay into the same cache.

(we need this fix to avoid issues into sale-workflow/sale_triple_discount)

acsonefho committed:
[10.0][FIX] Fix dependency issue

lmignon commented:
@acsonefho A little bit tricky to deal with the cache... it feels to me like a design problem into the addons or an issue into Odoo. #517 try to solve the same issue in an other way (the one with the cache not the triggers)

lmignon commented on account_invoice_triple_discount/models/account_invoice.py:
```diff
@@ -60,7 +67,7 @@ def triple_discount_preprocess(self):
                 discount3=line.discount3,
             )
-            line.update({
+            line._cache.update({
```
@acsonefho  Why do you deal with the cache? Is it required. Behind the scene it's what the `update`method do

acsonefho replied:
@lmignon I have to deal with the cache because (as the module is currently designed), every discount fields are computed into the standard discount field (into cache). Then, after calling the standard behaviour (`super`) the custom `discount` fields are re-updated with filled value.
But as there is `api.multi` and `api.one`, Odoo change the cache in some case.
I agree with your previous comment: maybe it's a design problem or an issue into Odoo core (so in this case, my fix is only a workaround).

SimoRubi commented on account_invoice_triple_discount/models/account_invoice.py:
```diff
@@ -32,8 +32,15 @@ class AccountInvoiceLine(models.Model):
+    # Each iteration of api.multi is done into a different cache!
+    # As we're using cache to define a value into discount field, we have
+    # to stay into the same cache. So we can not use api.multi in this case.
+    @api.one
```
Deprecated, see https://github.com/odoo/odoo/blob/feef39aee083294f8c8d19aecef0434af04c48d0/odoo/api.py#L359

acsonefho replied:
@SimoRubi Thank you for your comment but I explain why I was forced to use `@api.one` (instead of `multi`) on the line just before.

SimoRubi commented on account_invoice_triple_discount/models/account_invoice.py:
```diff
@@ -32,8 +32,15 @@ class AccountInvoiceLine(models.Model):
 
-    @api.multi
-    @api.depends('discount2', 'discount3')
+    # Each iteration of api.multi is done into a different cache!
```
I don't think this is the expected behavior, so looks like it is a bug of odoo core.
You might open an issue in there addressing the wrong behavior or create a PR if you have actually found the misbehaving code.

SimoRubi reviewed (changes requested):
Thanks for investigating this issue!
I already tried a different workaround in https://github.com/OCA/account-invoicing/pull/517

grindtildeath reviewed (approved):
Looks ok :+1:

rafaelbn commented:
Hi @SimoRubi @lmignon do you approve technically this PR?

SimoRubi commented:
> Hi @SimoRubi @lmignon do you approve technically this PR?

@rafaelbn as I explained in the review https://github.com/OCA/account-invoicing/pull/525#pullrequestreview-241586146 I'm actually torn from this because it re-introduces a deprecated decorator.
I'd like to know why it does not work and at least point it out (with an issue) to odoo core so it can hopefully fix it.
I can't do that because I didn't investigated enough and don't know the details, but maybe @acsonefho knows it better and could open an issue in odoo/odoo

rafaelbn commented:
Please @acsonefho @SimoRubi could we you clarify this in the Code Sprint?

eLBati reviewed (approved):
https://github.com/OCA/sale-workflow/issues/817#issuecomment-537047405

pedrobaeza commented:
@lmignon do you agree with this solution?

rousseldenis commented:
@lmignon

gurneyalex committed:
[FIX] account_invoice_triple_discount: cache can be reset

in some rare case, happening in a non deterministic way AFAICT, the cache can be reset,
resulting in wrong computations, because odoo will repopulate the cache with the data from the
database.

This patch attempts to detect this, and to fix the issue by relauching the method.

This is ugly, but in my opinion, messing with the cache as this module does is
a really bad idea.

gurneyalex commented:
Hello, we use this PR in production, and we experienced a bug when, in some rare random cases, the cache would be reset in the middle of the call to super()._compute_price(). It maybe because of some custom code for our customer which changes the dependencies. Still, it's not nice, because it shows that the code in this module is fragile (and we also had issues with sale_triple_discount for the same customer). 

This patch https://github.com/acsone/account-invoicing/pull/2 works around the issue for us by detecting the failure and re running the method. 

On the long term, I think that the approach of these modules should be changed by adding a discount1 field and making the standard discount field a computed field combining discount1 discount2 and discount3, and stopping to mess with the cache. 


cc @guewen @yvaucher

lmignon committed:
Merge pull request #2 from gurneyalex/PR525

[FIX] account_invoice_triple_discount: cache can be reset

pedrobaeza commented:
Let's merge finally this patch as the most satisfactory one:

/ocabot merge patch
